Mathlib/NumberTheory/EllipticDivisibilitySequence.lean

```diff
@@ -313,3 +834,418 @@ noncomputable def normEDSRec {P : ℕ → Sort u}
+
+section Complement
+
+variable (W₁ compl₂ : ℤ → R)
+  (h₁ : ∀ m, W 1 * W₁ m = W m) (h₂ : ∀ m, W m * compl₂ m = W (2 * m)) (m n : ℤ)
+
+/-- If `W` is an elliptic sequence whose first two terms are not zero divisors,
+the sequence constructed above indeed gives `W(n*m)` when multiplied by `W(m)`.
+The condition `mem` is actually redundant because `W` is a multiple of a normalised EDS
+by the other assumptions, so we can conclude using `normEDS_mul_compl` below. -/
+lemma IsEllSequence.mul_compl_eq_apply_mul_of_mem_nonZeroDivisors (mem : W m ∈ R⁰) :
+    W m * compl W₁ compl₂ m n = W (n * m) := by
+  induction n using Int.negInduction with
+  | nat n =>
+    refine n.strong_induction_on fun n ih ↦ ?_
+    cases' n with n; · simp [EllSequence.compl, ellW.zero 1 two]
+    cases' n with n; · simp [EllSequence.compl, compl']
+    rw [EllSequence.compl, Int.sign_eq_one_of_pos (by omega),
+      Int.natAbs_cast, compl', Int.cast_one, one_mul]
+    obtain ⟨k, rfl|rfl⟩ := n.even_or_odd'
+    · rw [dif_pos (even_two_mul _), k.mul_div_cancel_left zero_lt_two, mul_comm (compl₂ _),
+        ← mul_assoc, ← compl_ofNat, ih _ (by omega), h₂, ← mul_assoc, add_assoc, ← two_mul,
+        ← left_distrib, Nat.cast_mul]; rfl
+    simp_rw [dif_neg (Nat.not_even_two_mul_add_one _), show (2 * k + 1) / 2 = k by omega]
+    rw [← mul_cancel_right_mem_nonZeroDivisors (mul_mem mem <| pow_mem one 2)]
+    have := (ellW ((k + 1 + 1) * m) ((k + 1) * m) 1).symm
+    simp_rw [← right_distrib, ← mul_sub_right_distrib, add_sub_cancel_left,
+      ← h₁ (_ + 1), ← h₁ (_ - 1), ← Nat.cast_one (R := ℤ), ← Nat.cast_add] at this
+    rw [← ih _ (by omega), ← ih _ (by omega)] at this
+    simp_rw [compl_ofNat, Nat.cast_add] at this ⊢
+    convert this using 1
+    · ring_nf
+    rw [Nat.cast_mul]; ring_nf
+  | neg n hn => rw [neg_mul, ellW.neg one two, compl_neg, mul_neg, hn]
+
+lemma normEDS_mul_complEDS (m n : ℤ) :
+    normEDS b c d m * complEDS b c d m n = normEDS b c d (n * m) := by
+  obtain rfl|hm := eq_or_ne m 0; · simp
+  rw [normEDS_eq_aeval, universalNormEDS, complEDS_eq_aeval, ← map_mul]; congr 1
+  have := @universalNormEDS_mem_nonZeroDivisors
+  exact IsEllSequence.normEDS.mul_compl_eq_apply_mul_of_mem_nonZeroDivisors (this one_ne_zero)
+    (this two_ne_zero) _ _ (fun _ ↦ by simp) (fun _ ↦ normEDS_mul_compl₂EDS _ _ _ _) _ _ (this hm)
+
+lemma normEDS_mul_complEDS_div (hm : m ≠ 0) (n : ℤ) (dvd : m ∣ n) :
+    normEDS b c d m * complEDS b c d m (n / m) = normEDS b c d n := by
+  obtain ⟨n, rfl⟩ := dvd
+  rw [Int.mul_ediv_cancel_left _ hm, normEDS_mul_complEDS, mul_comm]
+
+namespace EllSequence
+
+variable (b c d)
+
+/-- An auxiliary expression that appears in the definition of the numerator of
+the reduced invariant and in the definition of the `ω` family of division polynomials. -/
+def invarNumAux : R :=
+  preNormEDS (b ^ 4) c d (m - 2) * preNormEDS (b ^ 4) c d (m + 1) ^ 2 * if Even m then 1 else b
```

Comment:
This file is getting long so I think I'll split everything about `invar, invarNumAux, redInvar` etc. into another file `Invar`. The `compl` stuff can't be split because they're used to show normEDS IsDivSequence.